### ISSUE TITLE
question about ranking loss

### ISSUE THREAD

**Issue by cv-lab-x** [Wanggcong/SparseNeRF#24]
hi, thanks for your work, some questions about the ranking loss code , as you mentioned in your paper, using dpt depth as supervised depth, if depth_a > depth_b do loss1 else do loss2, but in your loss codes, there're no comparision with dpt depth, how this works?      
`    margin1 = 1e-4
      margin2 = 1e-4
      depth_loss0_0 = jnp.mean(jnp.maximum(depth[0,:]-depth[1,:]+margin1,0)) ###
      depth_loss0_1 = jnp.mean(jnp.maximum(jnp.abs(depth[0,:]-depth[2,:])-margin2,0)) 
      depth_loss0_2 = jnp.mean(jnp.maximum(jnp.abs(depth[1,:]-depth[3,:])-margin2,0)) 
      depth_loss = depth_loss0_0+(depth_loss0_1+depth_loss0_2)*0.1
      all_depth_loss.append(depth_loss)`

looking forward to your reply! @Wanggcong

**Comment by Wanggcong** [comment]
> hi, thanks for your work, some questions about the ranking loss code , as you mentioned in your paper, using dpt depth as supervised depth, if depth_a > depth_b do loss1 else do loss2, but in your loss codes, there're no comparision with dpt depth, how this works? ` margin1 = 1e-4 margin2 = 1e-4 depth_loss0_0 = jnp.mean(jnp.maximum(depth[0,:]-depth[1,:]+margin1,0)) ### depth_loss0_1 = jnp.mean(jnp.maximum(jnp.abs(depth[0,:]-depth[2,:])-margin2,0)) depth_loss0_2 = jnp.mean(jnp.maximum(jnp.abs(depth[1,:]-depth[3,:])-margin2,0)) depth_loss = depth_loss0_0+(depth_loss0_1+depth_loss0_2)*0.1 all_depth_loss.append(depth_loss)`
> 
> looking forward to your reply! @Wanggcong

Thank you for your question. "jnp.mean(jnp.maximum(depth[0,:]-depth[1,:]+margin1,0))" is a marginal ranking loss. I invite you to refer to [this link](https://pytorch.org/docs/stable/generated/torch.nn.MarginRankingLoss.html). Please let me know if it answers your question.

**Comment by cv-lab-x** [comment]
> > hi, thanks for your work, some questions about the ranking loss code , as you mentioned in your paper, using dpt depth as supervised depth, if depth_a > depth_b do loss1 else do loss2, but in your loss codes, there're no comparision with dpt depth, how this works? ` margin1 = 1e-4 margin2 = 1e-4 depth_loss0_0 = jnp.mean(jnp.maximum(depth[0,:]-depth[1,:]+margin1,0)) ### depth_loss0_1 = jnp.mean(jnp.maximum(jnp.abs(depth[0,:]-depth[2,:])-margin2,0)) depth_loss0_2 = jnp.mean(jnp.maximum(jnp.abs(depth[1,:]-depth[3,:])-margin2,0)) depth_loss = depth_loss0_0+(depth_loss0_1+depth_loss0_2)*0.1 all_depth_loss.append(depth_loss)`
> > looking forward to your reply! @Wanggcong
> 
> Thank you for your question. "jnp.mean(jnp.maximum(depth[0,:]-depth[1,:]+margin1,0))" is a marginal ranking loss. I invite you to refer to [this link](https://pytorch.org/docs/stable/generated/torch.nn.MarginRankingLoss.html). Please let me know if it answers your question.


thanks for your reply, i mean in your paper, distiltion dpt depth rank to nerf, but didn't see any dpt monodepth in the loss
<img width="355" alt="image" src="https://github.com/Wanggcong/SparseNeRF/assets/26067407/859357d7-709a-4299-908a-de045c07bd6e">

`depth_loss0_0 = jnp.mean(jnp.maximum(depth[0,:]-depth[1,:]+margin1,0))`


<img width="428" alt="image" src="https://github.com/Wanggcong/SparseNeRF/assets/26067407/7f7fcbee-600d-4b21-9250-a4171183be36">

**Comment by Wanggcong** [comment]
Hi, we only use the relationship depth_a>depth_b or depth_a<=depth_b (so-called ranking in the paper). We do not use the depth values in the loss.

**Comment by massyzs** [comment]
Hi, I think this is really confusing:
Confusing part:
`
      margin1 = 1e-4
      margin2 = 1e-4
      depth_loss0_0 = jnp.mean(jnp.maximum(depth[0,:]-depth[1,:]+margin1,0)) ###
      depth_loss0_1 = jnp.mean(jnp.maximum(jnp.abs(depth[0,:]-depth[2,:])-margin2,0)) 
      depth_loss0_2 = jnp.mean(jnp.maximum(jnp.abs(depth[1,:]-depth[3,:])-margin2,0)) 
`
As you can see, you only make the abstraction for estimated depth and did nothing with DPT predicted one. I cannot understand your proposed two losses clearly.  **I even cannot find where did you return the dpt depth to main function.**

**Comment by Wanggcong** [comment]
> Hi, I think this is really confusing: Confusing part: `margin1 = 1e-4 margin2 = 1e-4 depth_loss0_0 = jnp.mean(jnp.maximum(depth[0,:]-depth[1,:]+margin1,0)) ### depth_loss0_1 = jnp.mean(jnp.maximum(jnp.abs(depth[0,:]-depth[2,:])-margin2,0)) depth_loss0_2 = jnp.mean(jnp.maximum(jnp.abs(depth[1,:]-depth[3,:])-margin2,0)) ` As you can see, you only make the abstraction for estimated depth and did nothing with DPT predicted one. I cannot understand your proposed two losses clearly. **I even cannot find where did you return the dpt depth to main function.**

Thank you for your question. I understand your question now. We sample the depth pixel values according to the dpt depth, please check this [code block](https://github.com/Wanggcong/SparseNeRF/blob/b545e5334f08999c51a6b0bb9cb8cc3767148f67/internal/datasets_depth_llff_dtu.py#L594C1-L604C44) . The "depth" variable in "depth_loss0_0 = jnp.mean(jnp.maximum(depth[0,:]-depth[1,:]+margin1,0))" in [code block](https://github.com/Wanggcong/SparseNeRF/blob/b545e5334f08999c51a6b0bb9cb8cc3767148f67/train_llff_dtu.py#L128C1-L132C87) actually has been ranked in [code block](https://github.com/Wanggcong/SparseNeRF/blob/b545e5334f08999c51a6b0bb9cb8cc3767148f67/internal/datasets_depth_llff_dtu.py#L594C1-L604C44).

Please let me know if I make it clear.

**Comment by cv-lab-x** [comment]
> > Hi, I think this is really confusing: Confusing part: `margin1 = 1e-4 margin2 = 1e-4 depth_loss0_0 = jnp.mean(jnp.maximum(depth[0,:]-depth[1,:]+margin1,0)) ### depth_loss0_1 = jnp.mean(jnp.maximum(jnp.abs(depth[0,:]-depth[2,:])-margin2,0)) depth_loss0_2 = jnp.mean(jnp.maximum(jnp.abs(depth[1,:]-depth[3,:])-margin2,0)) ` As you can see, you only make the abstraction for estimated depth and did nothing with DPT predicted one. I cannot understand your proposed two losses clearly. **I even cannot find where did you return the dpt depth to main function.**
> 
> @cv-lab-x @massyzs Thank you for your question. I understand your question now. We sample the depth pixel values according to the dpt depth, please check this [code block](https://github.com/Wanggcong/SparseNeRF/blob/b545e5334f08999c51a6b0bb9cb8cc3767148f67/internal/datasets_depth_llff_dtu.py#L594C1-L604C44) . The "depth" variable in "depth_loss0_0 = jnp.mean(jnp.maximum(depth[0,:]-depth[1,:]+margin1,0))" in [code block](https://github.com/Wanggcong/SparseNeRF/blob/b545e5334f08999c51a6b0bb9cb8cc3767148f67/train_llff_dtu.py#L128C1-L132C87) actually has been ranked in [code block](https://github.com/Wanggcong/SparseNeRF/blob/b545e5334f08999c51a6b0bb9cb8cc3767148f67/internal/datasets_depth_llff_dtu.py#L594C1-L604C44).
> 
> Please let me know if I make it clear.

yes,that's clear, thanks for your reply
